### ISSUE TITLE
Swing function.

### ISSUE THREAD

**Issue by reory** [LMMS/lmms#1091]
Hello new here.

I asked this question on the LMMS forum on facebook, it's not an issue but I was told to come here and sign up.

Is there any chance in the future of having a "swing" function on the bb editor?

http://blog.dubspot.com/swing-creative-use-of-groove-quantization/

Above link explains it better than I can. 

Great work by the way on the new LMMS beta and I love all the new sounds from zynaddsubfx. 

Thank you.

**Comment by zonkmachine** [comment]
I've actually implemented swing before but this was on a much simpler hardware sequencer. Since we've seen coding on the MIDI side of LMMS lately, probably the most sane way to go about it would be through groove templates.

Here's an interview with Roger Linn as a bonus   :)
http://www.attackmagazine.com/features/roger-linn-swing-groove-magic-mpc-timing/

**Comment by zonkmachine** [comment]
> probably the most sane way to go about it would be through groove templates.

I change that to: The easiest way is to have swing quantization in the Piano Roll editor. I don't think that would be picked up easily by the arpeggiator though but you wouldn't have to mess with the timing directly but just change the notes. I also had a look at the arpeggiator to see if it was feasible to make a swing knob in there but even if it's doable it would mess up the code pretty much. As I recall it that was the way it was done in Logic when I worked with it some 10+ years ago.

Basically a general algorithm for swing looks something like this:
frames, swing_factor, swing_frames1, swing_frames2
swing_factor = swingKnob * frames
swing_frames1 = frames + swing_factor
swing_frames2 = frames - swing_factor
step_in_beat % 2 ? swing_frames1 : swing_frames2

**Comment by teknopaul** [comment]
Hi I've actually done this. I had a version of LMMS locally with variable swing per track and a facility to add different groove functions as well as just simple swing.  So far just the swing algo talked about above, I'd need to port it to the new github project.  

My algo shifts the notes with more precision than just one 64th, to be honest shifting by 1 64th is pretty good, but I find 1 64th pretty limiting. (BTW technically in LMMS the limit is a 1 192th or a bar) 

If there is still demand for this I'll see if I can create a pull request with the changes.

Perhaps one day might be nice to import groves from other tools? I think there are some formats for transferring grooves.

**Comment by zonkmachine** [comment]
> If there is still demand for this I'll see if I can create a pull request with the changes.

Yes, yes, yeeees!!!

**Comment by teknopaul** [comment]
Done, https://github.com/LMMS/lmms/pull/3296 a bit rough round the edges but its a start.

**Comment by teknopaul** [comment]
I finished this, it supports save now, a simple UI and a couple of different groove functions as well as just Swing. I'm using it daily it seems pretty stable. I don't get how to submit the code though.

**Comment by BaraMGB** [comment]
Step 1: fork lmms on github. 
Step 2: clone your fork to your local computer 
Step 3: make a new branch (git checkout -b swing) 
Step 4: do your changes and test them. 
Step 5: git add . And git commit
Step 6: git push origin swing 
Step 7: do the pull request on github

**Comment by follower** [comment]
I've just randomly found @teknopaul's groove support in #3296 via a snippet of diff context in another issue.

Given the interest in this functionality and all @teknopaul's effort implementing it, it seems a shame to stumble at the last hurdle due to source control tooling (believe me, I'm sympathetic on that count :) ).

From looking at https://github.com/LMMS/lmms/pull/3296/commits it seems like (in comparison to https://github.com/teknopaul/lmms/commits/master) only the 3 required commits now appear in the PR?

The Qt5 Travis builds seem to fail due to this error ([e.g.](https://travis-ci.org/LMMS/lmms/jobs/195711670)):

    /Users/travis/build/LMMS/lmms/include/Groove.h:29:10: fatal error: 'QtGui/QWidget' file not found
    
    #include <QtGui/QWidget> 

Which seems to be related to a Qt 4 vs Qt 5 change ([see](https://stackoverflow.com/questions/22908977/qvtkwidget-cannot-include-qtgui-qwidget)).

I realise these things come down to available time but I thought I'd add this "current status" summary in case someone was inspired to take a look to get some cool functionality over the line (without, you know, having to implement the functionality themselves :D ) and/or to have some notes for myself if I get back to it first (unlikely in the near term due to how many yaks I'm in the middle of shaving currently :) ).

Thanks for your efforts getting it this far @teknopaul. :)

**Comment by tresf** [comment]
@follower thanks for resurrecting this.  The issue is that @teknopaul closed his PR before the feature was tested and merged.  I've reopened it for work to continue.  If you are comfortable working on this, we'd happily add you to the dev team to nudge it along.  Being on the dev team will give you access to push to his fork and get the feature implemented.

**Comment by tresf** [comment]
This statement is true for anyone that's willing to help us with PRs, BTW

**Comment by tresf** [comment]
@follower I've rebased #3296 against master, fixed build errors.

**Comment by rulatir** [comment]
"Swing" is a very limited model. We need to be able to warp the timing of events using different, pluggable mathematical models. The "delay every second note by a fixed amount" model is just one of dozens that various musical systems around the planet actually NEED.

https://github.com/sjaehn/BSchaffl/issues/19#issuecomment-1158026883

**Comment by teknopaul** [comment]
> "Swing" is a very limited model. We need to be able to warp the timing of events using different, pluggable mathematical models. The "delay every second note by a fixed amount" model is just one of dozens that various musical systems around the planet actually need.
> 
> [sjaehn/BSchaffl#19 (comment)](https://github.com/sjaehn/BSchaffl/issues/19#issuecomment-1158026883)

I think it always has been pluggable with different models

**Comment by teknopaul** [comment]
I have another implementation that is based on a simpler approach with  offsets in samples

https://github.com/teknopaul/lmms

This is very fast but does not survive large BPM changes very well.  Also this code does not have a global groove always applied per song, it has groove that can be assigned per note select Ctrl + G.

All the different "Swing" algorithms are re implemented.

Same approach could be used with offset as a float in fractions of a tick which would survive BPM changes.  I have not tried that but the same principal applies and each alog would need to be tweaked.

Both thesee approaches have the same feature that you can have different swing models in the same song.

The offset approach also allows for simple huamnization, and tuplets, flam and other tricks that require delays to not positioning.

![Screenshot from 2023-12-20 16-55-40](https://github.com/LMMS/lmms/assets/812666/d68ab0f4-7308-4ea7-8968-078cb1e31054)

**Comment by yohannd1** [comment]
noo, why was it closed?

**Comment by reory** [comment]
Over 10 years old. I've moved on. 😒

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by tresf** [comment]
Reopening.  Please feel free to unsubscribe if you wish.  Work started on this and is in WIP status here: https://github.com/LMMS/lmms/pull/4232.  Those interested in this  feature can attempt to rebase that branch against master, or open a new Pull Request.  This issue is still valid.

**Comment by reory** [comment]
oh wow. i thought it was over. ok reopen if you wish. thank you.
________________________________

**Comment by musikBear** [comment]
> oh wow. i thought it was over. ok reopen if you wish. thank you.

uHOh did you ever see my video on this...
https://www.youtube.com/watch?v=uX-LWhwxIQE&

It looks like i forgot to add a link here 3 years ago.. My bad....
